### PR TITLE
nlohmann_json_schema_validator: init at 2.3.0

### PR DESCRIPTION
**Pull request by br337**
## Description of changes

This package is a JSON schema validator for Modern C++. More information on it [here](https://github.com/pboettch/json-schema-validator). 

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your first contribution!

Things left to do:

1. Move your file in `pkgs/by-name/...` and remove the entry from `all-packages.nix`
2. The PR title should be: `<name>: init at <version>`
3. The commit message should be identical to the PR title
4. It would be nice to have you as maintainer, therefore a commit to add yourself in the maintainers file (`maintainers: add br337`) needs to be made before the current commit.
5. The `pname` looks wonky, why is it containing the name of one of its dependency ?

Let me know if you need some help!

**Comment by br337** [comment]
> Hi !
> 
> Thanks for your first contribution!
> 
> Things left to do:
> 
>     1. Move your file in `pkgs/by-name/...` and remove the entry from `all-packages.nix`
> 
>     2. The PR title should be: `<name>: init at <version>`
> 
>     3. The commit message should be identical to the PR title
> 
>     4. It would be nice to have you as maintainer, therefore a commit to add yourself in the maintainers file (`maintainers: add br337`) needs to be made before the current commit.
> 
>     5. The `pname` looks wonky, why is it containing the name of one of its dependency ?
> 
> 
> Let me know if you need some help!

Hi, I'm sorry if the following questions will seem kind of silly but I struggle with some things regarding this PR:

1. Why should this package go in `pkgs/by-name/` and not in `pkgs/development/libraries/` since it is a C++ dev library closely related to `nlohmann_json` which already resides in `pkgs/development/libraries/`?
2. `<name>` here being the name of the package and `<version>` the version of said library?
3. Noted! :smiley: 
4. Sure! That would be great. Would that require an additional PR or should that be done in this PR but just as an additional commit before the package has been committed? I assume I should add myself in the `/maintainers/maintainer-list.nix` file.
5. I know it's wonky, but I'm not the original developer of the library and that is sadly the name the library uses pretty much everywhere, including in CMake when importing it.

**Comment by drupol** [comment]
> Hi, I'm sorry if the following questions will seem kind of silly but I struggle with some things regarding this PR:
> 
>     1. Why should this package go in `pkgs/by-name/` and not in `pkgs/development/libraries/` since it is a C++ dev library closely related to `nlohmann_json` which already resides in `pkgs/development/libraries/`?

This is the new package name convention ([RFC140](https://github.com/NixOS/rfcs/pull/140)). Maybe you should have a look and see if there are already libraries in there. If yes, then I would kindly ask you to follow that pattern.

>     2. `<name>` here being the name of the package and `<version>` the version of said library?

Correct.

>     4. Sure! That would be great. Would that require an additional PR or should that be done in this PR but just as an additional commit before the package has been committed? I assume I should add myself in the `/maintainers/maintainer-list.nix` file.

Yes, you can make a commit within this PR with you added in the maintainers list.

>     5. I know it's wonky, but I'm not the original developer of the library and that is sadly the name the library uses pretty much everywhere, including in CMake when importing it.

Fair enough, since I don't know anything in there, I'll trust you in there.

File: maintainers/maintainer-list.nix
Comment (drupol) [inline maintainers/maintainer-list.nix:2619]: Sorry to be picky, but can you please sort attributes alphabetically? Just like most of other entries?

File: pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix
Comment (drupol) [inline pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:26]: Same here please, sort the attributes alphabetically.

**Review by drupol — CHANGES_REQUESTED** [review]

File: maintainers/maintainer-list.nix
Comment (br337) [reply maintainers/maintainer-list.nix]: Sure thing.

But keep in mind that the top comment in `maintainers-list.nix` isn't sorted in the correct order either. https://github.com/NixOS/nixpkgs/blob/4bfea513e208c4175fae8b59c0f94233b508c958/maintainers/maintainer-list.nix#L3-L16

The ordering should also be correct now

File: pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix
Comment (br337) [reply pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix]: This should be in the correct order now.

File: maintainers/maintainer-list.nix
Comment (drupol) [reply maintainers/maintainer-list.nix]: This is something I might fix! Thanks for noticing it. That said, sorting attributes alphabetically is something common in Nix, so, the good habit is to sort them alphabetically.

File: pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix
Comment (drupol) [inline pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:14]:
```suggestion
    rev = finalAttrs.version;
```

File: pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix
Comment (drupol) [inline pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:27]:
```suggestion
  meta = {
    description = "JSON schema validator for JSON for Modern C++ ";
    homepage = "https://github.com/pboettch/json-schema-validator";
    license = lib.licenses.mit;
    maintainers = with lib.maintainers; [ br337 ];
    platforms = platforms.all;
  };
```

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix
Comment (drupol) [reply pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix]: lib.platforms.all ! I forgot it !

File: pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix
Comment (drupol) [inline pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:26]:
```suggestion
    platforms = lib.platforms.all;
```

**Comment by br337** [comment]
Should I also squash the last commits?

**Comment by drupol** [comment]
> Should I also squash the last commits?

Correct.

**Comment by br337** [comment]
> > Should I also squash the last commits?
> 
> Correct.

Done

**Review by drupol — APPROVED** [review]
Congrats for your first PR, welcome to the family !
It LGTM, but let's wait for someone else's review.

Keep PRs coming!

File: pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix
Comment (pbsds) [inline pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:10]:
```suggestion

  outputs = [ "out" "dev" ];

```

**Review by pbsds — APPROVED** [review]
Result of `nixpkgs-review pr 275224` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nlohmann_json_schema_validator</li>
  </ul>
</details>

Runs fine, LGTM. Just a nit

File: pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix
Comment (drupol) [reply pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:10]: I don't even understand this yet and how to learn how to know we have to do that :)

File: pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix
Comment (2xsaiko) [inline pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:22]:
```suggestion
    description = "JSON schema validator for JSON for Modern C++";
```

**Review by dblsaiko — CHANGES_REQUESTED** [review]
Just a small thing :^)

**Comment by 2xsaiko** [comment]
1 package built:
nlohmann_json_schema_validator

Result of `nixpkgs-review pr 275224` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>nlohmann_json_schema_validator</li>
  </ul>
</details>

**Comment by br337** [comment]
Changes are now squashed

**Review by drupol — APPROVED** [review]

**Review by dblsaiko — APPROVED** [review]

File: pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix
Comment (Janik-Haag) [inline pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:29]: please add meta.mainProgram, since the pr adds multiple binarys and things like `lib.getExe` or `nix run` won't work without it.

Comment (pbsds) [reply pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:29]: Unless there is a clear _single_ binary the package provides which is the one users expect to run, i oppose setting `mainProgram`. In such cases one is expected to use `lib.getExe'` or `nix shell PKG -c binary ...`.

Comment (pbsds) [reply pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:29]: Being able to disable the `bin/${pname}` fallback behavior of `lib.getExe` with something like `mainProgram = null` would be great, but implementing that that is outside the scope of this PR

Comment (Janik-Haag) [reply pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:29]: > Unless there is a clear binary the package provides which is the one users expect to be run, i oppose setting mainProgram. In such cases one is expected to use lib.getExe' or nix shell PKG -c binary ....

I would have set it to `json-schema-validate` since that's the thing in the package name. But feel free to merge if you think otherwise, the package looks good to me except for this.

Comment (br337) [reply pkgs/by-name/nl/nlohmann_json_schema_validator/package.nix:29]: So should I implement this change now or should I resolve the issue without further changes?

**Comment by drupol** [comment]
Let's not wait too long before getting feedback, especially on a first contributor PR.
If anything need to be changed now, it can be done either by the author or someone else.

Thanks for your contribution, keep PRs coming!

**Pull request merged**
